### PR TITLE
Fsdp prefetch params

### PR DESCRIPTION
**Pull request by ngoyal2707**
Option to prefetch previous layer's parameters in zero3 configuration. This allows to have `ReduceScatter` overlap with backward computation.

* Baseline profile: 
Screen Shot 2021-11-17 at 5 37 07 PM
The `ReduceScatter` and `AllGather` are overlapped with backward computation in zero3 config.

* Prefetch during post backward profile:
Screen Shot 2021-11-17 at 5 37 15 PM
`ReduceScatter` are perfectly overlapped with backward computation. (this gives around 20% speed up for the model I benchmarked) 

I also tried Prefetch during pre backward profile but that doesn't seem to help further, this still needs a bit more debugging

### CODE REVIEW

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -274,6 +274,9 @@ def __init__(
         force_input_to_fp32: bool = False,
         verbose: bool = False,
         cpu_offload: bool = False,
+        gradient_predivide_factor: Optional[float] = None,
```
Comment (anj-s) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:277]: Do you need to sync to main?

We are moving away from adding params to the API. can you create a config object that is suitable for this keeping in mind other params that can be added as well?

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -1062,6 +1070,8 @@ def _set_is_root(self) -> None:
         # As the root, we now set all children instances to False and
         # give them a closure to try to queue a wait_for_post_backward.
         self.children_share_process_group = True
+        
+        # fsdp_module_list = []
```
Comment (anj-s) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1074]: remove?

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -936,8 +941,11 @@ def summon_full_params(self, recurse: bool = True, volatile: bool = False) -> Ge
     def _reset_lazy_init(self) -> None:
         """Reset instance so :func:`_lazy_init` will run on the next forward."""
         self._is_root: Optional[bool] = None
+        self._fsdp_forward_ordering: List[nn.Module] = []
```
Comment (anj-s) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:944]: can you add a comment about what this is tracking?

**Comment by anj-s** [comment]
@ngoyal2707  thank you for the PR! can you add a PR description of the changes made to the FSDP code? Can you also add unit tests to make sure these changes work as expected?

cc @tmarkstrum  who was looking into overlapping the reduce-scatter and allgather streams with the BW computation.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -1345,6 +1370,15 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
...
             return
 
+        if (
+            self.reshard_after_forward
+            and self.enable_prefetch_next_module_params
+            and not self.prefetch_before_backward
+            and self._fsdp_forward_ordering is not None 
+            and self._my_fsdp_instance_idx is not None and self._my_fsdp_instance_idx > 0
+        ):
+            self._fsdp_forward_ordering[self._my_fsdp_instance_idx - 1]._rebuild_full_params()
```
Comment (sshleifer) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1380]: here we want to wait for all_gather_stream?

**Review by sshleifer — APPROVED** [review]
Nice! 
Consider adding some coverage to one of the files in `tests/nn/data_parallel/`.

**Comment by min-xu-ai** [comment]
Thanks for the PR!

Tingting might have a different method that address this overlapping more fully. I'd suggest we can wait a bit to see what's her results look like. What do you think?

**Comment by tmarkstrum** [comment]
any test cases here?

**Comment by anj-s** [comment]
@ngoyal2707 Do we still want this change in FairScale? cc @tmarkstrum

**Comment by anj-s** [comment]
We've gone with a different approach of using multiple process group. Closing this PR for now.

**Pull request closed without merging**